### PR TITLE
Improve error message on server when a vanilla client connects

### PR DESCRIPTION
**Pull request by mezz**
When a vanilla client connection to a Forge server fails, this lists the mods that are incompatible with vanilla.

This helps the situation here, where someone wants to create a Forge server that is compatible with vanilla clients:
https://www.reddit.com/r/feedthebeast/comments/7rukv2/not_sure_why_vanilla_users_have_to_download_forge/

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java [hunk 1]
```diff
@@ -635,8 +636,9 @@ public boolean handleServerHandshake(C00Handshake packet, NetworkManager manager
         if (packet.getRequestedState() == EnumConnectionState.LOGIN && (!NetworkRegistry.INSTANCE.isVanillaAccepted(Side.CLIENT) && !packet.hasFMLMarker()))
         {
             manager.setConnectionState(EnumConnectionState.LOGIN);
-            TextComponentString text = new TextComponentString("This server requires FML/Forge to be installed. Contact your server admin for more details.");
-            FMLLog.log.info("Disconnecting Player: {}", text.getUnformattedText());
+            TextComponentString text = new TextComponentString("This server has mods that require FML/Forge to be installed on the client. Contact your server admin for more details.");
+            Collection<String> modNames = NetworkRegistry.INSTANCE.getModNamesThatDoNotAcceptVanilla(Side.CLIENT);
+            FMLLog.log.info("Disconnecting Player: This server has mods that require FML/Forge to be installed on the client:\n{}", modNames);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: I'd put a tab after the new line to indicate the messages go together. But looks fine to me. Except your method name is pretty big, could just be 'getRequiredMods(SIDE)'

File: src/main/java/net/minecraftforge/fml/common/network/NetworkRegistry.java [hunk 2]
```diff
@@ -304,13 +306,18 @@ public void register(ModContainer fmlModContainer, Class<?> clazz, @Nullable Str
...
-            result &= e.getValue().acceptsVanilla(from);
-        }
-        return result;
+        return registry.values().stream()
+                .allMatch(mod -> mod.acceptsVanilla(from));
     }
+
+    public Collection<String> getModNamesThatDoNotAcceptVanilla(Side from)
+    {
+        return registry.values().stream()
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/fml/common/network/NetworkRegistry.java]: Might be nice to have the list here sorted?

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java (hunk 1 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: Sounds good, will do.

File: src/main/java/net/minecraftforge/fml/common/network/NetworkRegistry.java (hunk 2 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/network/NetworkRegistry.java]: Yeah, it's not performance-critical or anything so might as well make it nice. Will do.

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java (hunk 1 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: I'll just put it on one line, it's going to be really long either way and this way it'll all be together.

**Pull request merged**